From ab93bc4e925ec731e26cfdfe142be519f648f887 Mon Sep 17 00:00:00 2001
From: Ente <ducksource@duckpond.ch>
Date: Tue, 13 Apr 2021 22:55:40 +0200
Subject: [PATCH] Send HTTP credentials when fetching blobs

With this change media-tag now sends HTTP credentials when fetching
blobs. Also changed the example nginx config to send
Access-Control-Allow-Credentials CORS headers. For this to work, we can
no longer use '*' for Access-Control-Allow-Origin [1][2]: Therefore the
example config was changed to set Access-Control-Allow-Origin to the
sandbox domain only.

Fixes:
- #705: Blob fetch fails with 401 Unauthorized when HTTP basic auth is enabled [3]

Referenes:
[1]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
[2]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials
[3]: https://github.com/xwiki-labs/cryptpad/issues/705
---
 docs/example.nginx.conf | 9 ++++++---
 www/common/media-tag.js | 4 ++++
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/docs/example.nginx.conf b/docs/example.nginx.conf
index 2c677436b..fe3bd7eaf 100644
--- a/docs/example.nginx.conf
+++ b/docs/example.nginx.conf
@@ -57,7 +57,8 @@ server {
     add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
     add_header X-XSS-Protection "1; mode=block";
     add_header X-Content-Type-Options nosniff;
-    add_header Access-Control-Allow-Origin "*";
+    add_header Access-Control-Allow-Origin https://$sandbox_domain;
+    add_header Access-Control-Allow-Credentials true;
     # add_header X-Frame-Options "SAMEORIGIN";
 
     set $coop '';
@@ -169,7 +170,8 @@ server {
     # encrypted blobs are immutable and are thus cached for a year
     location ^~ /blob/ {
         if ($request_method = 'OPTIONS') {
-            add_header 'Access-Control-Allow-Origin' '*';
+            add_header 'Access-Control-Allow-Origin' https://$sandbox_domain;
+            add_header 'Access-Control-Allow-Credentials' true;
             add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
             add_header 'Access-Control-Allow-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range';
             add_header 'Access-Control-Max-Age' 1728000;
@@ -178,7 +180,8 @@ server {
             return 204;
         }
         add_header Cache-Control max-age=31536000;
-        add_header 'Access-Control-Allow-Origin' '*';
+        add_header 'Access-Control-Allow-Origin' https://$sandbox_domain;
+        add_header 'Access-Control-Allow-Credentials' true;
         add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
         add_header 'Access-Control-Allow-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range,Content-Length';
         add_header 'Access-Control-Expose-Headers' 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range,Content-Length';
diff --git a/www/common/media-tag.js b/www/common/media-tag.js
index 15b038724..7763f4966 100644
--- a/www/common/media-tag.js
+++ b/www/common/media-tag.js
@@ -234,6 +234,7 @@ var factory = function () {
         var check = function () {
             var xhr = new XMLHttpRequest();
             xhr.open("HEAD", src);
+            xhr.withCredentials = true;
             xhr.onerror = function () { return void cb("XHR_ERROR"); };
             xhr.onreadystatechange = function() {
                 if (this.readyState === this.DONE) {
@@ -266,6 +267,7 @@ var factory = function () {
         var fetch = function () {
             var xhr = new XMLHttpRequest();
             xhr.open('GET', src, true);
+            xhr.withCredentials = true;
             xhr.responseType = 'arraybuffer';
 
             var progress = function (offset) {
@@ -377,6 +379,7 @@ var factory = function () {
         var fetch = function () {
             var xhr = new XMLHttpRequest();
             xhr.open('GET', src, true);
+            xhr.withCredentials = true;
             xhr.setRequestHeader('Range', 'bytes=0-1');
             xhr.responseType = 'arraybuffer';
 
@@ -389,6 +392,7 @@ var factory = function () {
                 var xhr2 = new XMLHttpRequest();
 
                 xhr2.open("GET", src, true);
+                xhr2.withCredentials = true;
                 xhr2.setRequestHeader('Range', 'bytes=2-' + (size + 2));
                 xhr2.responseType = 'arraybuffer';
                 xhr2.onload = function () {
-- 
2.30.1

